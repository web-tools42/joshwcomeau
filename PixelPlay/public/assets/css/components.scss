@import 'variables';

.card {
  background: #EEE;
  border-bottom: 1px solid rgba(0,0,0,0.8);
  border-right: 1px solid rgba(0,0,0,0.8);

  &:before {
    content: " ";
    position: absolute;
    z-index: -1;
    top: -$cardBorder;
    left: -$cardBorder;
    right: -$cardBorder;
    bottom: -$cardBorder;
    border: $cardBorder solid #BBB;
  }

}