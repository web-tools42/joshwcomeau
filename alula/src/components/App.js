import React, {Component} from 'react';
import {Route} from 'react-router-dom'
import {Provider} from 'react-redux';
import createHistory from 'history/createBrowserHistory'

import { ConnectedRouter } from 'react-router-redux'
import configureStore from '../store';

import Intro from './Intro';
import Main from './Main';
import DevTools from './DevTools';
import {FullHeight} from './utility-components';


const history = createHistory();

const store = configureStore(history);

class App extends Component {
  componentDidMount() {
    // On some mobile devices, tapping twice causes a zoom-in.
    // We want to disable this, as it's super annoying when trying to undo
    // multiple steps.
    document.addEventListener('touchend', event => {
      // Make sure that tap events don't bubble up to the window.
      event.stopPropagation();
    });
  }


  render() {
    return (
      <Provider store={store}>
        <ConnectedRouter history={history}>
          <FullHeight>
            <Route exact path="/" component={Intro} />
            <Route path="/create" component={Main} />

            <DevTools />
          </FullHeight>
        </ConnectedRouter>
      </Provider>
    );
  }
}

export default App;
