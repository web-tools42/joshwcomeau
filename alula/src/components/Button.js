import React, {Component} from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import {colors, media, styles} from '../constants';


const Button = styled.button`
  position: relative;
  height: ${styles.buttonHeightPx};
  width: ${props => props.width};
  padding: 0;
  border-radius: 2px;
  border-width: ${props => props.fill ? '0px' : '2px'};
  border-style: solid;
  border-color: ${props => props.fill ? 'none' : props.borderColor};
  background: ${props => props.fill ? props.color : 'rgba(0,0,0,0.5)'};
  color: ${props => props.fill ? colors.white : props.color};
  font-size: 18px;
  font-weight: 400;
  outline: none;
  cursor: pointer;

  &:disabled {
    opacity: 0.5;
  }

  ${media.xs`
    height: ${styles.buttonHeightXSPx};
  `}
`;

Button.defaultProps = {
  width: 'auto',
  color: colors.grays[2],
  borderColor: colors.grays[3],
  fill: false,
}

export default Button;
