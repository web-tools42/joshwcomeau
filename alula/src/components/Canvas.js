import React, {PureComponent} from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import {
  applyTransformation,
  undoTransformation,
  rightClickCanvas,
} from '../actions';
import {getCurrentCanvas} from '../reducers/history.reducer';
import {disableEventOnMobile} from '../utils/event.utils';
import {
  scaleCanvas,
  getPixelRatio,
  getCursorPosition,
  getCroppedImageParams,
  mirrorTransformLine,
} from '../utils/canvas.utils';


const CanvasElem = styled.canvas`
  position: relative;
  z-index: 1;
  display: block;
  width: 100%;
  height: 100%;
  background: #CCC;
`;

class Canvas extends PureComponent {
  static propTypes = {
    image: PropTypes.object,
  }

  componentDidMount() {
    // Make the image a square, based on the largest dimension (either
    // full-width or full-height)
    const size = Math.min(window.innerWidth, window.innerHeight);

    this.canvas.width = size;
    this.canvas.height = size;

    this.pixelRatio = getPixelRatio(this.ctx);

    this.ctx.imageSmoothingEnabled = false;

    scaleCanvas(this.canvas, this.ctx);

    if (this.props.image) {
      this.updateImage(this.props.image);
      this.props.applyTransformation(this.canvas);
    }
  }

  componentWillReceiveProps(nextProps) {
    const newImageSelected = nextProps.image !== this.props.image;
    const newTransformationReceived = nextProps.canvas !== this.props.canvas;

    if (newImageSelected) {
      this.updateImage(nextProps.image);
      this.props.applyTransformation(this.canvas);
    }

    if (newTransformationReceived) {
      this.updateImage(nextProps.canvas);
    }
  }

  updateImage = (image, callback) => {
    const canvasWidth = this.canvas.width / this.pixelRatio;
    const canvasHeight = this.canvas.height / this.pixelRatio;

    const croppedImageParams = getCroppedImageParams({
      canvasWidth,
      canvasHeight,
      imageWidth: image.width,
      imageHeight: image.height,
      pixelRatio: this.pixelRatio,
    });

    this.ctx.drawImage(
      image,
      ...croppedImageParams
    );
  }

  getEventCoords = (ev) => {
    // This method normalizes the difference between touch events and mouse
    // events, to return a set of X/Y coordinates for an event regardess
    // of input device.
    const coordHolder = ev.touches ? ev.touches[0] : ev;

    try {
      getCursorPosition(coordHolder, this.canvas)
    } catch (e) {
      alert(e);
    }

    return getCursorPosition(coordHolder, this.canvas);
  }

  startDrag = (ev) => {
    disableEventOnMobile(ev);

    // If the user right-clicks, we don't want to do any mirroring.
    // We _do_ want to signal this intent, though, since there's a good
    // chance it means they're done and are saving the image.
    if (ev.button === 2) {
      this.props.rightClickCanvas();
      return;
    }

    const {x: x1, y: y1} = this.getEventCoords(ev);
    this.x1 = x1;
    this.y1 = y1;

    this.isDragging = true;
  }

  handleDrag = (ev) => {
    if (!this.isDragging) {
      return;
    }

    // At the start of each move event, restore the canvas to the previously
    // saved state. This is necessary because each move updates the canvas
    // as a 'preview'. The state isn't saved until the drag is released.
    this.ctx.drawImage(
      this.props.canvas,
      0,
      0,
      this.canvas.width / this.pixelRatio,
      this.canvas.height / this.pixelRatio,
    );

    this.ctx.save();


    const {x1, y1} = this;
    const {x: x2, y: y2} = this.getEventCoords(ev);

    const sideA = x2 - x1;
    const sideB = y2 - y1;

    let fullLine;

    // If our line is perfectly vertical, the standard `ax + b` form won't
    // work. In this case, though, our job is easy.
    if (sideA === 0) {
      const x = x1;

      fullLine = {
        x: x1 * this.pixelRatio,
        y1: 0,
        y2: this.canvas.height,
      };

      this.ctx.setTransform(-1, 0, 0, 1, x * this.pixelRatio * 2, 0)

      this.ctx.beginPath();
      this.ctx.moveTo(fullLine.x, 0);
      this.ctx.lineTo(fullLine.x, this.canvas.height);
      this.ctx.lineTo(this.canvas.width, this.canvas.height);
      this.ctx.lineTo(this.canvas.width, 0);
    } else {
      // Extend our line so that it reaches the edge of the canvas.
      // Given that this is just a straight line, it can be represented by
      // the simple equation `y = ax + b`.
      // We can figure out the slope, `a`, pretty easily.
      const slope = sideB / sideA;
      // We can work out the offset `b` by solving the equation for one of our
      // known points.
      // y1 = slope * x1 + offset
      const offset = y1 - (slope * x1);

      // Now that we have this data, extend our line to span the entire canvas.
      fullLine = {
        x1: 0 * this.pixelRatio,
        y1: offset * this.pixelRatio,
        x2: this.canvas.width * this.pixelRatio,
        y2: (slope * this.canvas.width + offset) * this.pixelRatio,
      };

      this.ctx.setTransform(...mirrorTransformLine(fullLine));

      this.ctx.beginPath();
      this.ctx.moveTo(fullLine.x1, fullLine.y1);
      this.ctx.lineTo(fullLine.x2, fullLine.y2);
      this.ctx.lineTo(this.canvas.width, 0);
      this.ctx.lineTo(0, 0);
    }

    this.ctx.clip();

    // draw the image
    // since the this.ctx is rotated, the image will be rotated also
    this.ctx.drawImage(
      this.canvas,
      0,
      0,
    );

    // weâ€™re done with the rotating so restore the unrotated this.ctx
    this.ctx.setTransform(1,0,0,1,0,0);

    this.ctx.restore();
  }

  releaseDrag = (ev) => {
    disableEventOnMobile(ev);

    this.isDragging = false;

    this.props.applyTransformation(this.canvas);
    // this.ctx.translate(this.canvas.width / 2, this.canvas.height / 2);
    // this.ctx.rotate(true ? -Math.PI / 2 : Math.PI / 2);

  }

  storeRefToCanvas = (canvas) => {
    // If the component unmounts before the ref can be stored,
    // bail early.
    if (!canvas) {
      return;
    }

    this.canvas = canvas;
    this.ctx = canvas.getContext("2d");
  }

  render() {
    return (
      <CanvasElem
        innerRef={this.storeRefToCanvas}
        onMouseDown={this.startDrag}
        onMouseMove={this.handleDrag}
        onMouseUp={this.releaseDrag}
        onTouchStart={this.startDrag}
        onTouchMove={this.handleDrag}
        onTouchEnd={this.releaseDrag}
      />
    );
  }
}

const mapStateToProps = state => ({
  image: state.image,
  canvas: getCurrentCanvas(state),
});

const mapDispatchToProps = {
  applyTransformation,
  undoTransformation,
  rightClickCanvas,
};

export default connect(mapStateToProps, mapDispatchToProps)(Canvas);
