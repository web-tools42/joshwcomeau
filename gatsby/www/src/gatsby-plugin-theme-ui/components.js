import React from "react"

import GuideList from "../components/guide-list.js"
import HubspotForm from "../components/hubspot-form"
import Pullquote from "../components/shared/pullquote"
import EggheadEmbed from "../components/shared/egghead-embed"
import DateChart from "../components/chart"
import CodeBlock from "../components/code-block"

export default {
  GuideList,
  HubspotForm,
  DateChart,
  Pullquote,
  EggheadEmbed,
  pre: ({ children }) => <CodeBlock>{children}</CodeBlock>,
}
