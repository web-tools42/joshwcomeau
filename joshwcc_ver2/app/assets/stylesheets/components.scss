h1 {
  font-size: 4rem;
  letter-spacing: -1px;
  font-weight: 900;
}

h2 {
  font-size: 2.5rem;
  font-weight: 700;
}

h3 {
  font-size: 2.25rem;
  font-weight: 500;
  letter-spacing: -1px;
}

h4 {
  font-size: 2rem;
  font-weight: 500;
}

h5 {
  font-size: 1.75rem;
  font-weight: 500;
}