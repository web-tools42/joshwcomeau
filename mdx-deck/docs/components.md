
# Components

mdx-deck includes a few built-in components to help with creating presentations.

## Head

Use the `<Head />` component to set content in the document head.

```mdx
// example for twitter cards
import { Head } from 'mdx-deck'

<Head>
  <title>My Presentation</title>
  <meta name='twitter:card' content='summary_large_image' />
  <meta name='twitter:site' content='@jxnblk' />
  <meta name='twitter:title' content='My Presentation' />
  <meta name='twitter:description' content='A really great presentation' />
  <meta name='twitter:image' content='https://example.com/card.png' />
</Head>
```

## Image

Use the `<Image />` component to render a fullscreen image (using the CSS `background-image` property).

```mdx
import { Image } from 'mdx-deck'

<Image src='kitten.png' />
```

### Props
- `src` (string) image URL
- `size` (string) CSS background-size

## Appear

Use the `<Appear />` component to make its children appear one at a time within a single slide.
Use the left and right arrow keys to step through each element.

```mdx
import { Appear } from 'mdx-deck'

<ul>
  <Appear>
    <li>One</li>
    <li>Two</li>
    <li>Three</li>
  </Appear>
</ul>
```

## Speaker Notes

Speaker notes that only show in presenter mode can be added to any slide with either markdown syntax or with the Notes component.

````mdx
# Markdown speaker notes

```notes
These are only visible in presenter mode
```
````

```mdx
import { Notes } from 'mdx-deck'

# Slide Content

<Notes>
  Only visible in presenter mode
</Notes>
```

## Layouts

mdx-deck includes a few built-in layouts for common slide variations.
Export a layout as the `default` within a slide to wrap the contents.

### Invert

Inverts the foreground and background colors from the theme.

```mdx
import { Invert } from 'mdx-deck/layouts'

# Normal

---

export default Invert

# Inverted
```

### Split

Creates a horizontal layout with the first child on the left and all other children on the right.

```mdx
import { Split } from 'mdx-deck/layouts'

export default Split

![](kitten.png)

## Meow
```

### SplitRight

Same as the Split component, but renders the first child on the right.

```mdx
import { SplitRight } from 'mdx-deck/layouts'

export default SplitRight

![](kitten.png)

## Meow
```

### FullScreenCode

Render fenced code blocks fullscreen.

````mdx
import { FullScreenCode } from 'mdx-deck/layouts'

export default FullScreenCode

```jsx
<Button>Beep</Button>
```
````

