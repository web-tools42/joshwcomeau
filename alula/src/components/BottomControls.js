import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import styled from 'styled-components';
import RotateLeftIcon from 'react-icons/lib/md/rotate-left';
import RotateRightIcon from 'react-icons/lib/md/rotate-right';
import UndoIcon from 'react-icons/lib/md/undo';
import RestoreIcon from 'react-icons/lib/md/delete';

import {colors, media, styles} from '../constants';
import {getCanUndo} from '../reducers/history.reducer';
import {
  undoTransformation,
  rotateCW,
  rotateCCW,
  restoreOriginalImage,
} from '../actions';

import Button from './Button';
import {Row, IconAdjustment} from './utility-components';


const Wrapper = styled.div`
  position: relative;
  z-index: 2;
  padding: ${styles.paddingUnitPx};
  background: ${styles.backgroundColor};
  color: ${styles.textColor};
  display: flex;
  flex-direction: column;
  border-top: 1px solid rgba(255,255,255,0.05);
`;

class BottomControls extends PureComponent {
  static propTypes = {
    canUndo: PropTypes.bool.isRequired,
    undoTransformation: PropTypes.func.isRequired,
  }

  handleRestore = () => {
    const msg = 'This will reset the image to its original state! Are you sure?';

    // eslint-disable-next-line no-restricted-globals
    if (confirm(msg)) {
      this.props.restoreOriginalImage();
    }
  }

  render() {
    const {
      canUndo,
      undoTransformation,
      rotateCW,
      rotateCCW,
    } = this.props;

    return (
      <Wrapper>
        <Row>
          <Button
            width={styles.buttonHeightPx}
            color={colors.reds[2]}
            borderColor={colors.reds[3]}
            onClick={this.handleRestore}
          >
            <IconAdjustment><RestoreIcon /></IconAdjustment>
          </Button>

          <Button
            width={styles.buttonHeightPx}
            color={colors.pinks[1]}
            borderColor={colors.pinks[2]}
            onClick={rotateCCW}
            style={{ marginLeft: styles.paddingUnit }}
          >
            <IconAdjustment><RotateLeftIcon /></IconAdjustment>
          </Button>
          <Button
            width={styles.buttonHeightPx}
            color={colors.pinks[1]}
            borderColor={colors.pinks[2]}
            onClick={rotateCW}
            style={{ marginLeft: styles.paddingUnit }}
          >
            <IconAdjustment><RotateRightIcon /></IconAdjustment>
          </Button>

          <Button
            color={colors.purples[1]}
            borderColor={colors.purples[2]}
            onClick={undoTransformation}
            disabled={!canUndo}
            style={{ flex: 1, marginLeft: styles.paddingUnit }}
          >
            <IconAdjustment><UndoIcon /></IconAdjustment>
          </Button>
        </Row>
      </Wrapper>
    )
  }
}

const mapStateToProps = state => ({
  canUndo: getCanUndo(state),
});

const mapDispatchToProps = {
  undoTransformation,
  rotateCW,
  rotateCCW,
  restoreOriginalImage,
};

export default connect(mapStateToProps, mapDispatchToProps)(BottomControls);
