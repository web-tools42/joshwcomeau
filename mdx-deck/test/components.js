import React from 'react'
import { create as render } from 'react-test-renderer'
import 'jest-styled-components'
import { ThemeProvider } from 'styled-components'
import components from '../src/components'

const blacklist = {
  pre: true,
  img: true
}

test.each(
  Object.keys(components)
    .filter(key => !blacklist[key])
    .map(key => {
      const Component = components[key]
      return [ key, <ThemeProvider theme={{}}><Component /></ThemeProvider> ]
    })
)('renders %s', (name, el) => {
  const json = render(el).toJSON()
  expect(json).toMatchSnapshot()
})
