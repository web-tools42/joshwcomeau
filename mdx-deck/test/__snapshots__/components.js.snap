// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a 1`] = `
.c0 {
  color: link;
}

<a
  className="c0"
  color="link"
/>
`;

exports[`renders blockquote 1`] = `
.c0 {
  text-align: left;
  font-weight: bold;
  font-size: 16px;
  margin-left: 0px;
  margin-right: 0px;
  padding-left: 0px;
  padding-right: 0px;
  color: quote;
}

<blockquote
  className="c0"
  color="quote"
  fontSize={2}
/>
`;

exports[`renders code 1`] = `
.c0 {
  text-align: left;
  white-space: pre-wrap;
  font-size: 14px;
  margin: 0px;
  padding: 8px;
  color: pre;
  background-color: preBackground;
}

<pre
  className="c0"
  color="pre"
  fontSize={1}
/>
`;

exports[`renders h1 1`] = `
.c0 {
  line-height: 1.25;
  font-size: 24px;
  margin-bottom: 16px;
  margin-top: 16px;
  color: heading;
}

<h1
  className="c0"
  color="heading"
  fontSize={4}
/>
`;

exports[`renders h2 1`] = `
.c0 {
  line-height: 1.25;
  font-size: 20px;
  margin-bottom: 16px;
  margin-top: 16px;
  color: heading;
}

<h2
  className="c0"
  color="heading"
  fontSize={3}
/>
`;

exports[`renders h3 1`] = `
.c0 {
  line-height: 1.25;
  font-size: 16px;
  margin-bottom: 16px;
  margin-top: 16px;
  color: heading;
}

<h3
  className="c0"
  color="heading"
  fontSize={2}
/>
`;

exports[`renders h4 1`] = `
<h4
  className=""
  fontSize={2}
/>
`;

exports[`renders h5 1`] = `
<h5
  className=""
  fontSize={2}
/>
`;

exports[`renders h6 1`] = `
<h6
  className=""
  fontSize={2}
/>
`;

exports[`renders inlineCode 1`] = `
.c0 {
  color: code;
  background-color: codeBackground;
}

<code
  className="c0"
  color="code"
/>
`;

exports[`renders li 1`] = `
<li
  className=""
/>
`;

exports[`renders ol 1`] = `
.c0 {
  text-align: left;
  font-size: 16px;
}

<ol
  className="c0"
  fontSize={2}
/>
`;

exports[`renders p 1`] = `
.c0 {
  font-size: 16px;
}

<p
  className="c0"
  fontSize={2}
/>
`;

exports[`renders table 1`] = `
.c0 {
  overflow-x: auto;
}

.c1 {
  width: 100%;
  border-collapse: separate;
  border-spacing: 0;
}

.c1 td,
.c1 th {
  text-align: left;
  padding-right: .5em;
  padding-top: .25em;
  padding-bottom: .25em;
  border-bottom: 1px solid;
  vertical-align: top;
}

<div
  className="c0"
>
  <table
    className="c1"
  />
</div>
`;

exports[`renders ul 1`] = `
.c0 {
  text-align: left;
  font-size: 16px;
}

<ul
  className="c0"
  fontSize={2}
/>
`;
